samwise.js

```diff
@@ -1,58 +1,56 @@
 /**
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?